File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l = len(heights)

        if l == 0:
            return 0

        i = 0
        j = l - 1
        if heights[i] > heights[j]:
            maxSquare = (j - i) * heights[j]
        else:
            maxSquare = (j - i) * heights[i]

        while i < j:
            if heights[i] > heights[j]:
                j -= 1
            else:
                i += 1

            if heights[i] > heights[j]:
                newSquare = (j - i) * heights[j]
            else:
                newSquare = (j - i) * heights[i]

            if maxSquare < newSquare:
                maxSquare = newSquare    

        return maxSquare

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<ArrayList<Integer>> stack = new Stack<ArrayList<Integer>>();
        int maxSquare = 0;

        for (int i = 0; i <= heights.length; i++) {
            int h = 0;
            if (i < heights.length) {
                h = heights[i];
            }

            int p = 0;
            int l = 0;
            for (;;) {
                if (stack.empty()) {
                    break;
                } else {
                    ArrayList<Integer> top = stack.peek();

                    if (top.get(0) > h) {
                        stack.pop();

                        l += (top.get(1) + 1);
                        int square = l * top.get(0);
                        if (square > maxSquare) {
                            maxSquare = square;
                        }

                        p++;
                    } else {
                        break;
                    }
                }
            }


            ArrayList<Integer> val = new ArrayList<Integer>();
            val.add(h);
            val.add(l);

            stack.push(val);
        }

        return maxSquare;
    }
}
